Rewrite of the Secure Websocket explanation (#4223)
w3f/polkadot-wiki: docs/maintain/maintain-wss.md

A typically setting for an externally accessible polkadot archive RPC node would be:

```config
polkadot --chain polkadot --name myrpc --state-pruning archive --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944
```

Or for a polkadot pruned RPC node:

```config
polkadot --chain polkadot --name myrpc --state-pruning 1000 --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944
```

The specified flag option are outlined in greater detail below.

### Archive Node vs Pruned Node

A pruned node only keeps a limited number of finalized blocks of the network and not its full
history. Most frequently required actions can be completed with a pruned node, such as displaying
account balances, making transfers, setting up session keys, staking, etc. An archive node has the
full history (database) of the network and can be queried in all kind of ways, such as providing
historical information regarding transfers, balance histories, and more advanced queries involving
past events, etc.

An archive node requires a lot more diskspace. Start of januari 2023 a pruned polkadot takes around

> 115 GB, a polkadot archive node around 765 GB, which will inrease slowly over time. For an archive

Reviewer: Minor nitpick on wording:
> Start of januari 2023 a pruned polkadot takes around
115 GB, a polkadot archive node around 765 GB, which will inrease slowly over time

> At the start of January 2023, polkadot disk usage was at 115 GB for a pruned node and 765 GB for an archive node. This value will increase with time.

Author: Changed

---
id: maintain-wss
title: Secure WebSocket
sidebar_label: Secure WebSocket
description: Steps on setting up a secure socket for remote connections.
keywords: [web socket, remote, connection, secure websocket]
slug: ../maintain-wss
---

> A websocket provides access to the RPC server part of a substrate node, which can be used as a way

Reviewer: > A websocket provides access to the RPC server part of a substrate node

To be precise, I would phrase it as: `The substrate node RPC server can be accessed over the websocket protocol`

to access the underlying network and/or validator node. By default you can access your node's RPC
server from localhost (for example to rotate keys or do other maintenance) but to access it from
from another server or from an applications UI (such as [Polkadot-JS UI](https://polkadot.js.org/apps)),

> you will need to enable a secure proxy for the 

Reviewer: If think it's important to point that you don't **absolutely need** a secure proxy. The "secure proxy" is recommended for enabling access to the RPC node over an TLS connection and as such encrypting the connection between the end user and RPC server. Also many browsers such as Google Chrome will block non secure WS endpoints if they come from a different origin.

websocket connection.

:::note Enabling remote access to your validator node should not be necessary and is not
suggested as it can often lead to security problems :::

## Set up a node

Setting up any Substrate-based node relies on a similar process.  For example, they will all by default share the same websocket connection at port 9944 on localhost.
In this example, we'll set up a Kusama sync node on a debian flavoured server (such as Ubuntu 22.04). Create a new server on your provider of choice or locally at home. See
[Set up a Full Node](./maintain-sync) for additional instructions. You can choose to install from the default
apt repository or build from scratch. The startup options  in the setup process provide a variety of 
different settings that can be modified.

A typically setting for an externally accessible kusama archive RPC node would be:

```config
polkadot --chain kusama --name myrpc --state-pruning archive --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944
```

Or for a kusama pruned RPC node:

```config

> polkadot --chain kusama --name myrpc --state-pruning 1000 --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944

Reviewer: The example is good, however I would replace references to Kusama by Polkadot as this is the polkadot docs after all :smiley:.

Author: Changed to polkadot. Can you confirm that `--state-pruning 1000 --blocks-pruning archive` is correct for a pruned node or should it just be `--state-pruning 1000` without the blocks-pruning?

```

The specified flag option are outlined in greater detail below.

### Archive Node vs Pruned Node

> A pruned node only tracks more recent information about the network and not its full history. Most

Reviewer: Replace `only tracks more recent information about the network`  by `only keeps a limited number of finalized blocks of the network`

frequently required actions can be completed with a pruned node, such as displaying account balances, making
transfers, setting up session keys, staking, etc. An archive node has the full history (database) of the
network and can be queried in all kind of ways, such as providing historical information regarding transfers, balance histories, and more advanced queries involving past events, etc.

> An archive node requires a lot more diskspace ([overview](https://stakeworld.io/snapshot)). For an

Reviewer: > [overview](https://stakeworld.io/snapshot)

Maybe let's make clear that it is an external link. For example, using a sentence like `According to statistics provided by Stakework, on Jan 6th 2023 a polkadot prune node was 100GB.`

---
id: maintain-wss
title: Secure WebSocket
sidebar_label: Secure WebSocket
description: Steps on setting up a secure socket for remote connections.
keywords: [web socket, remote, connection, secure websocket]
slug: ../maintain-wss
---

A websocket provides access to the RPC server part of a substrate node, which can be used as a way
to access the underlying network and/or validator node. By default you can access your node's RPC
server from localhost (for example to rotate keys or do other maintenance) but to access it from

> from another server or for example connect to it from on UI like
> [Polkadot-JS UI](https://polkadot.js.org/apps) you will need to enable a secure proxy for the

Reviewer 2: ```suggestion
from another server or from an applications UI (such as [Polkadot-JS UI](https://polkadot.js.org/apps)),
you will need to enable a secure proxy for the 
```

websocket connection.

> :::note Enabling remote access to your validator node should not be neccessary and can only lead to
> security problems :::

Reviewer 2: ```suggestion
:::note Enabling remote access to your validator node should not be necessary and is not
suggested as it can often lead to security problems :::
```

## Set up a node

> For this example, we'll set up a Kusama sync node on a debian flavoured server (Ubuntu 22.04 for
> example). Create a new server on your provider of choice or locally at home. See
> [Set up a Full Node](./maintain-sync) for instructions, you can choose to install from the default
> apt repository of build from scratch, whatever you prefer. In the startup options you can choose
> different settings.

Reviewer 2: ```suggestion
Setting up any Substrate-based node relies on a similar process.  For example, they will all by default share the same websocket connection at port 9944 on localhost.
In this example, we'll set up a Kusama sync node on a debian flavoured server (such as Ubuntu 22.04). Create a new server on your provider of choice or locally at home. See
[Set up a Full Node](./maintain-sync) for additional instructions. You can choose to install from the default
apt repository or build from scratch. The startup options  in the setup process provide a variety of 
different settings that can be modified.
```

A typically setting for an externally accessible kusama archive RPC node would be:

```config
polkadot --chain kusama --name myrpc --state-pruning archive --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944
```

Or for a kusama pruned RPC node:

```config
polkadot --chain kusama --name myrpc --state-pruning 1000 --blocks-pruning archive --ws-max-connections 100 --rpc-cors all --rpc-methods Safe --ws-port 9944
```

> See below for explanation about the options.

Reviewer 2: ```suggestion
The specified flag option are outlined in greater detail below.
```

> ### Archive node vs pruned node
> 
> A pruned node knows only the recent information about the network and not its full history. Most
> frequently done actions can be done with a pruned node, for example see account balances, make
> transfers, setup session keys, staking, etc. An archive node has the full history (database) of the
> network and can be queried in all kind of ways, give information about transfers since the network
> started, historical balances, advanced queries about past events, etc.

Reviewer 2: ```suggestion
### Archive Node vs Pruned Node

A pruned node only tracks more recent information about the network and not its full history. Most
frequently required actions can be completed with a pruned node, such as displaying account balances, making
transfers, setting up session keys, staking, etc. An archive node has the full history (database) of the
network and can be queried in all kind of ways, such as providing historical information regarding transfers, balance histories, and more advanced queries involving past events, etc.
```

An archive node requires a lot more diskspace ([overview](https://stakeworld.io/snapshot)). For an
archive node you need the options `--state-pruning archive --blocks-pruning archive` in your startup
settings.

:::tip Inclusion in the Polkadot.js UI requires an archive node. :::

### Secure the RPC server

> Via the node startup settings you can choose **what** to expose with **how many** connections **from
> where** through your rpc server.

Reviewer 2: ```suggestion
The node startup settings allow you to choose **what** to expose, **how many** connections to expose and **from
where** access should be granted through the rpc server.
```

_How many_: You can set your maximum connections through `--ws-max-connections`, for example
`--ws-max-connections 100`

_From where_: by default localhost and the polkadot.js are allowed to access the RPC server, you can
change this by setting `--rpc-cors`, to allow access from everywhere you need `--rpc-cors all`

_What_: you can limit the methods to use with `--rpc-methods`, an easy way to set this to a safe
mode is `--rpc-methods Safe`

## Secure the ws port

> The main idea is converting the non secure ws port to a secure wss port by putting it behind a ssl
> enabled proxy. So from outside one see's the ssl enabled apache2/nginx/other proxy server, which
> redirects the request to the internal rpc node. For this you will need an SSL certificate. There are
> different possible approaches; an official certificate like letsencrypt or using a self signed
> certificate.

Reviewer 2: ```suggestion
A non secure ws port can be converted to a secure wss port by placing it behind an SSL
enabled proxy. The SSL enabled apache2/nginx/other proxy server is not externally visible, which
redirects requests to the internal rpc node. For this you will need an SSL certificate. There are
different strategies for obtaining a cert, such as using a service like letsencrypt or self signing.
```

> ### Get a SSL certificate

Reviewer 2: ```suggestion
### Obtaining an SSL Certificate
```

> Easiest is to get a real (free) SSL certificate following LetsEncrypt instructions

Reviewer 2: ```suggestion
One easy way to get a free SSL certificate can be achieved by following the LetsEncrypt instructions
```

([nginx](https://certbot.eff.org/instructions?ws=nginx&os=ubuntufocal)/[apache](https://certbot.eff.org/instructions?ws=apache&os=ubuntufocal)).
This will auto-generate an SSL certificate and include it in your configuration.

Alternatively you can generate a self-signed certificate and rely on the raw IP address of your node
when connecting to it. This is not preferable since you will have to whitelist the certificate to
access it from a browser.

```bash
sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout /etc/ssl/private/selfsigned.key -out /etc/ssl/certs/selfsigned.crt
sudo openssl dhparam -out /etc/ssl/certs/dhparam.pem 2048
```

> ## Install a proxy server

Reviewer 2: ```suggestion
## Installing a Proxy Server
```

}
```

### Apache2

> Apache2 is slightly more computationally intensive when compared to nginx but also provides more configurable options. You can run it in different modes such as prefork, worker or event. In this example, we use

Reviewer: > Apache2 is slightly more computationally intensive when compared to nginx but also provides more configurable options

I think we should refrain from doing any judgement on Apache vs nginx but ideally provide the 2 config examples.

Here is a snipped based on our internal config. Could you kindly double check it on your side and add it.

```
server {
  listen                443 ssl;
  server_name           YOUR_DOMAIN;
# for letsencrypt
  ssl_certificate /etc/letsencrypt/live/YOUR_DOMAIN/fullchain.pem;
  ssl_certificate_key /etc/letsencrypt/live/YOUR_DOMAIN/privkey.pem;
# for self-signed
  ssl_certificate /etc/ssl/certs/selfsigned.crt;
  ssl_certificate_key /etc/ssl/private/selfsigned.key;
  location / {
    limit_req zone=zone burst=5;
    proxy_buffers 16 4k;
    proxy_buffer_size 2k;
    proxy_pass ws://localhost:9944;
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header Host $host;
   }
}
```

Author: I made some more changes; I find the very specific ssl instruction making it more confusing. I think that somebody who sets up a rpc node will know how to setup a ssl enabled host and the specific rpc/proxypass config is the relevant part. Let me know if you disagree and think it needs to be more explicit.

}
```

### Apache2

> Apache2 is a little heavier then nginx but also has some more tweaking posibilities. You can run it
> in different modes, prefork, worker or event. For this manual we use
> [event](https://httpd.apache.org/docs/2.4/mod/event.html) which works well on higher load
> enviroments but other modes also work fine.

Reviewer 2: ```suggestion
Apache2 is slightly more computationally intensive when compared to nginx but also provides more configurable options. You can run it in different modes such as prefork, worker or event. In this example, we use
[event](https://httpd.apache.org/docs/2.4/mod/event.html) which works well on higher load
environments but other modes are also useful given the requirements.
```

```bash
apt install apache2
a2dismod mpm_prefork
a2enmod mpm_event proxy proxy_html proxy_http proxy_wstunnel rewrite ssl
```

The [mod_proxy_wstunnel](https://httpd.apache.org/docs/2.4/mod/mod_proxy_wstunnel.html) provides
_support for the tunnelling of web socket connections to a backend websockets server. The connection

> is automatically upgraded to a websocket connection_. In a ssl enabled virtualhost add:

Reviewer 2: ```suggestion
is automatically upgraded to a websocket connection_. In a SSL enabled virtualhost add:
```

```apacheconf
SSLProxyEngine on
ProxyRequests off

ProxyPass / ws://localhost:9944
ProxyPassReverse / ws://localhost:9944
```

> ## Connect to the node

Reviewer 2: ```suggestion
## Connecting to the Node
```
